physical/raft/fsm.go

```diff
@@ -590,34 +591,66 @@ func (f *FSM) ApplyBatch(logs []*raft.Log) []interface{} {
 	// We will construct one slice per log, each slice containing another slice of results from our get ops
 	entrySlices := make([][]*FSMEntry, 0, numLogs)
 
+	// We're also going to return errors
+	entryErrors := make([]error, 0, numLogs)
+	numErrors := 0
+
+	createResponse := func(success bool, entrySlices [][]*FSMEntry, entryErrors []error) []interface{} {
+		resp := make([]interface{}, 0, numLogs)
+		for i := range logs {
+			fsmar := &FSMApplyResponse{
+				Success: success,
+			}
+
+			if success {
+				fsmar.EntrySlice = entrySlices[i]
+			} else {
+				if len(entryErrors) == numLogs {
+					fsmar.Error = entryErrors[i]
+				} else {
+					fsmar.Error = entryErrors[0]
+				}
+			}
+
+			resp = append(resp, fsmar)
+		}
+
+		return resp
+	}
+
 	// Do the unmarshalling first so we don't hold locks
 	var latestConfiguration *ConfigurationValue
 	commands := make([]interface{}, 0, numLogs)
-	for _, log := range logs {
-		switch log.Type {
+	for _, l := range logs {
+		switch l.Type {
 		case raft.LogCommand:
 			command := &LogData{}
-			err := proto.Unmarshal(log.Data, command)
+			err := proto.Unmarshal(l.Data, command)
 			if err != nil {
 				f.logger.Error("error proto unmarshaling log data", "error", err)
-				panic("error proto unmarshaling log data")
+				entryErrors = append(entryErrors, fmt.Errorf("error proto unmarshaling log data: %w", err))
+				numErrors++
 			}
 			commands = append(commands, command)
 		case raft.LogConfiguration:
-			configuration := raft.DecodeConfiguration(log.Data)
-			config := raftConfigurationToProtoConfiguration(log.Index, configuration)
-
+			configuration := raft.DecodeConfiguration(l.Data)
+			config := raftConfigurationToProtoConfiguration(l.Index, configuration)
 			commands = append(commands, config)
 
 			// Update the latest configuration the fsm has received; we will
 			// store this after it has been committed to storage.
 			latestConfiguration = config
-
 		default:
-			panic(fmt.Sprintf("got unexpected log type: %d", log.Type))
+			entryErrors = append(entryErrors, fmt.Errorf("got unexpected log type: %d", l.Type))
```

Comment:
Isn't returning an error going to have similar results to the panic?  In other words, I think we should just ignore or log warnings for unknown operation types.

Reply:
Ok